### ISSUE TITLE
InvalidCastException with async, MediaType & MediaFoundationEncoder.Encode

### ISSUE THREAD

**Issue by BrainCrumbz** [naudio/NAudio#39]
Hello all

we'd just like to notify a problem we encountered. Not sure if we are doing it wrong or the library should behave differently, but thought to post it here in case other people find the same.

This bit of async method works

````c#
public async Task<int> JoinAsync(string outputPathName, IEnumerable<MediaFoundationReader> waveReaders)
{
  await Task.Run(() =>
  {
    MediaType wmaMediaType = CreateMediaType();  // calls MediaFoundationEncoder.SelectMediaType with some constants and enums

    using (CustomJoiningWaveProvider joiningWaveProvider = new CustomJoiningWaveProvider(waveReaders))
    using (MediaFoundationEncoder wmaEncoder = new MediaFoundationEncoder(wmaMediaType))
    {
      // encoder does not offer an async interface, so just wrap that to offload work on some thread
      wmaEncoder.Encode(outputPathName, joiningWaveProvider);
    }
  });
  // ...
}
````

*CustomJoiningWaveProvider* implements *IWaveProvider* and is able to join many *MediaFoundationReader*s (from Wav files) into a *MemoryStream*.

If we move the *MediaType* creation outside of *Task.Run* (as well as moving out the *using*s too) like in the following:

````c#
public async Task<int> JoinAsync(string outputPathName, IEnumerable<MediaFoundationReader> waveReaders)
{
  MediaType wmaMediaType = CreateMediaType();  // calls MediaFoundationEncoder.SelectMediaType with some constants and enums

  await Task.Run(() =>
  {
    using (CustomJoiningWaveProvider joiningWaveProvider = new CustomJoiningWaveProvider(waveReaders))
    using (MediaFoundationEncoder wmaEncoder = new MediaFoundationEncoder(wmaMediaType))
    {
      // encoder does not offer an async interface, so just wrap that to offload work on some thread
      wmaEncoder.Encode(outputPathName, joiningWaveProvider);
    }
  });
  // ...
}
````

the `wmaEncoder.Encode` call causes a `System.InvalidCastException` exception:

> Message: Specified cast is not valid

> StackTrace:
   in System.StubHelpers.InterfaceMarshaler.ConvertToNative(Object objSrc, IntPtr itfMT, IntPtr classMT, Int32 flags)
   in NAudio.MediaFoundation.IMFSinkWriter.AddStream(IMFMediaType pTargetMediaType, Int32& pdwStreamIndex)
   in NAudio.Wave.MediaFoundationEncoder.Encode(String outputFile, IWaveProvider inputProvider)
   in <--- our async method --->
   in System.Threading.Tasks.Task.InnerInvoke()
   in System.Threading.Tasks.Task.Execute()

> HRESULT: 0x80004002

Maybe that's related to COM interoperability, not sure. Apart from using the working version of code, should we do something more? TA

**Comment by markheath** [comment]
yeah threading is a bit of a nightmare with these COM objects. Best to use them on the same thread that created them. There is a post-compiling IL hack that can supposedly fix this type of thing, which the SharpDX project use and I'd like to eventually move into NAudio.

**Comment by BrainCrumbz** [comment]
Thanks for reply. What seems surprising is that even just the creation of the *MediaType* instance in a thread, and using that in a different thread, causes problems. 
Somehow it would seem more *comprehensible* if the problem arose from creating the encoder in a thread and then using that in another thread. But the *MediaType* seemed a simpler, so-to-say static piece of data that would cause no harm.
